### PR TITLE
Refactor WalDeltaData

### PR DESCRIPTION
**Pull request by Ignition**
It was one big struct, replaced with variant.

### CODE REVIEW

**Review by as51340 — APPROVED** [review]
Nice work!

File: src/storage/v2/durability/wal.cpp [hunk 1]
```diff
@@ -142,50 +143,281 @@ constexpr Marker DeltaActionToMarker(Delta::Action action) {
...
+    return *std::move(str);
+  } else {
+    if (!decoder->SkipPropertyValue()) throw RecoveryFailure("Invalid WAL data!");
+  }
+}
+
+template <bool is_read, auto VER>
+auto Decode(utils::tag_type<VersionDependantGid<VER>> /*version*/, BaseDecoder *decoder, const uint64_t version)
+    -> std::conditional_t<is_read, std::optional<Gid>, void> {
+  if (version <= VER) {
```
Comment (Ignition) [inline src/storage/v2/durability/wal.cpp:273]: Good job for tests...it caught this edge case. I've got my inequality the wrong away around.
Only if `version` is `VER` or greater do we do the read/skip.

File: src/storage/v2/durability/wal.cpp [hunk 2]
```diff
@@ -142,50 +143,281 @@ constexpr Marker DeltaActionToMarker(Delta::Action action) {
...
+  }
+}
+
+template <bool is_read>
+auto Decode(utils::tag_type<std::string> /*version*/, BaseDecoder *decoder, const uint64_t /*version*/)
+    -> std::conditional_t<is_read, std::string, void> {
+  if constexpr (is_read) {
+    auto str = decoder->ReadString();
+    if (!str) throw RecoveryFailure("Invalid WAL data!");
+    return *std::move(str);
```
[the remark this answers is not in the document]
Comment (Ignition) [reply src/storage/v2/durability/wal.cpp:252]: identical code-gen https://godbolt.org/z/nzYcbjncG
